File: lifeos/reviews/LIFEOS-P3-143/independent-re-review-1/review_tools/verify_lineage.py
#!/usr/bin/env python3
"""Reviewer-owned non-self-referential hash verifier for the supplied P3-143 commit."""

import hashlib
import json
import os
import sys
from pathlib import Path

WORKSPACE = Path.cwd()
CANDIDATE_COMMIT = "767301fb051d11a6bf11488413f8b03aabfc2d99"
MANIFEST = WORKSPACE / "lifeos/engineering/LIFEOS-P3-143/FINAL_MANIFEST.json"
CLOSURE = WORKSPACE / "lifeos/engineering/LIFEOS-P3-143/evidence/root_authority_closure.json"
OUTPUT = WORKSPACE / "lifeos/reviews/LIFEOS-P3-143/independent-re-review-1/evidence/lineage_verification.json"


def digest(payload: bytes) -> str:
    return hashlib.sha256(payload).hexdigest()


def main() -> int:
    manifest = json.loads(MANIFEST.read_text())
    records = []
    category_counts = {category: 0 for category in manifest["categories"]}
    for entry in manifest["files"]:
        absolute = (MANIFEST.parent / entry["path"]).resolve()
        payload = absolute.read_bytes()
        actual_hash = digest(payload)
        item = {
            "category": entry["category"],
            "role": entry.get("role"),
            "path": os.path.relpath(absolute, WORKSPACE),
            "bytes_expected": entry["bytes"],
            "bytes_actual": len(payload),
            "sha256_expected": entry["sha256"],
            "sha256_actual": actual_hash,
            "pass": len(payload) == entry["bytes"] and actual_hash == entry["sha256"],
        }
        records.append(item)
        category_counts[item["category"]] = category_counts.get(item["category"], 0) + 1
    closure_payload = CLOSURE.read_bytes()
    closure = json.loads(closure_payload)
    errors = sum(not record["pass"] for record in records)
    output = {
        "schema": "lifeos.p3-143.independent-lineage-verification.v1",
        "reviewer_owned": True,
        "candidate_commit": CANDIDATE_COMMIT,
        "manifest_self_referential": manifest["self_referential"],
        "manifest_entry_count": len(records),
        "category_counts": category_counts,
        "hash_error_count": errors,
        "root_authority_closure": {
            "sha256": digest(closure_payload),
            "candidate_files": closure["candidate_identity"]["candidate_files"],
            "closure_candidate_sha256": closure["candidate_identity"]["after_root_authority_closure_sha256"],
            "decision": closure["decision"],
        },
        "status": "PASS" if not manifest["self_referential"] and errors == 0 else "FAIL",
        "files": records,
    }
    OUTPUT.write_text(json.dumps(output, ensure_ascii=False, indent=2) + "\n")
    return 0 if output["status"] == "PASS" else 1


if __name__ == "__main__":
    sys.exit(main())

File: lifeos/reviews/LIFEOS-P3-143/independent-re-review-1/review_tools/verify_lineage.mjs
import { createHash } from "node:crypto";
import { readFile, writeFile } from "node:fs/promises";
import { dirname, relative, resolve } from "node:path";

const workspace = process.cwd();
const candidateCommit = "767301fb051d11a6bf11488413f8b03aabfc2d99";
const manifestPath = resolve(workspace, "lifeos/engineering/LIFEOS-P3-143/FINAL_MANIFEST.json");
const closurePath = resolve(workspace, "lifeos/engineering/LIFEOS-P3-143/evidence/root_authority_closure.json");
const outputPath = resolve(workspace, "lifeos/reviews/LIFEOS-P3-143/independent-re-review-1/evidence/lineage_verification.json");

const sha256 = (bytes) => createHash("sha256").update(bytes).digest("hex");
const manifest = JSON.parse(await readFile(manifestPath, "utf8"));
const results = [];
for (const entry of manifest.files) {
  const absolute = resolve(dirname(manifestPath), entry.path);
  const relativePath = relative(workspace, absolute);
  const bytes = await readFile(absolute);
  const actual = sha256(bytes);
  results.push({
    category: entry.category,
    role: entry.role ?? null,
    path: relativePath,
    bytes_expected: entry.bytes,
    bytes_actual: bytes.length,
    sha256_expected: entry.sha256,
    sha256_actual: actual,
    pass: bytes.length === entry.bytes && actual === entry.sha256,
  });
}
const closureBytes = await readFile(closurePath);
const closure = JSON.parse(closureBytes);
const categories = Object.fromEntries(manifest.categories.map((category) => [category, 0]));
for (const item of results) categories[item.category] = (categories[item.category] ?? 0) + 1;
const errors = results.filter((item) => !item.pass).length;
const output = {
  schema: "lifeos.p3-143.independent-lineage-verification.v1",
  reviewer_owned: true,
  candidate_commit: candidateCommit,
  manifest_self_referential: manifest.self_referential,
  manifest_entry_count: results.length,
  category_counts: categories,
  hash_error_count: errors,
  root_authority_closure: {
    sha256: sha256(closureBytes),
    candidate_files: closure.candidate_identity?.candidate_files ?? null,
    closure_candidate_sha256: closure.candidate_identity?.after_root_authority_closure_sha256 ?? null,
    decision: closure.decision ?? null,
  },
  status: !manifest.self_referential && errors === 0 ? "PASS" : "FAIL",
  files: results,
};
await writeFile(outputPath, `${JSON.stringify(output, null, 2)}\n`, { mode: 0o600 });
if (output.status !== "PASS") process.exitCode = 1;
